src/frequenz/client/reporting/_client.py

```diff
@@ -0,0 +1,289 @@
+# License: MIT
+# Copyright © 2024 Frequenz Energy-as-a-Service GmbH
+
+"""Client for requests to the Reporting API."""
+
+from collections import namedtuple
+from dataclasses import dataclass
+from datetime import datetime
+from typing import Any, AsyncIterator, Awaitable, Generator, Type, cast
+
+import grpc.aio as grpcaio
+
+# pylint: disable=no-name-in-module
+from frequenz.api.common.v1.metrics.metric_sample_pb2 import Metric as PBMetric
+from frequenz.api.common.v1.microgrid.microgrid_pb2 import (
+    MicrogridComponentIDs as PBMicrogridComponentIDs,
+)
+from frequenz.api.common.v1.pagination.pagination_params_pb2 import (
+    PaginationParams as PBPaginationParams,
+)
+from frequenz.api.reporting.v1.reporting_pb2 import (
+    ListMicrogridComponentsDataRequest as PBListMicrogridComponentsDataRequest,
+)
+from frequenz.api.reporting.v1.reporting_pb2 import (
+    ListMicrogridComponentsDataResponse as PBListMicrogridComponentsDataResponse,
+)
+from frequenz.api.reporting.v1.reporting_pb2 import TimeFilter as PBTimeFilter
+from frequenz.api.reporting.v1.reporting_pb2_grpc import ReportingStub
+from frequenz.client.common.metric import Metric
+from google.protobuf.timestamp_pb2 import Timestamp as PBTimestamp
+
+# pylint: enable=no-name-in-module
+
+Sample = namedtuple("Sample", ["timestamp", "value"])
+"""Type for a sample of a time series."""
+
+MetricSample = namedtuple(
+    "MetricSample", ["timestamp", "microgrid_id", "component_id", "metric", "value"]
+)
+"""Type for a sample of a time series incl. metric type, microgrid and component ID"""
+
+
+@dataclass(frozen=True)
+class ComponentsDataPage:
+    """A page of microgrid components data returned by the Reporting service."""
+
+    _data_pb: PBListMicrogridComponentsDataResponse
+    """The underlying protobuf message."""
+
+    def is_empty(self) -> bool:
+        """Check if the page contains valid data.
+
+        Returns:
+            True if the page contains no valid data.
+        """
+        if not self._data_pb.microgrids:
+            return True
+        if not self._data_pb.microgrids[0].components:
+            return True
+        if not self._data_pb.microgrids[0].components[0].metric_samples:
+            return True
+        return False
+
+    def iterate_metric_samples(self) -> Generator[MetricSample, None, None]:
```

Comment:
I would just make the class iterable directly (also `Iterator[MetricSample]`[^1] is a shorter and more idiomatic alias for `Generator[MetricSample, None, None]`[^2])

```py
    def __iter__(self) -> Iterator[MetricSample]:
        ...
```

Then you can just do `for sample in page:` instead of `for sample in page.iterate_metric_samples():`.

[^1]: Technically a [`Generator`](https://peps.python.org/pep-0342/) is more than an `Iterator`:

    > extends [iterators](https://docs.python.org/3.11/glossary.html#term-iterator) with the [send()](https://docs.python.org/3.11/reference/expressions.html#generator.send), [throw()](https://docs.python.org/3.11/reference/expressions.html#generator.throw) and [close()](https://docs.python.org/3.11/reference/expressions.html#generator.close) methods

    But in this context I think we only care about the iterator part, we are using a generator out of convenience, not because we are actually using the extra methods provided by a generator.

[^2]: From [`typing.Generator`](https://docs.python.org/3.11/library/typing.html#typing.Generator):

    > Alternatively, annotate your generator as having a return type of either Iterable[YieldType] or Iterator[YieldType]:

Reply:
This is a small change, so we can also briefly discuss here and could be quickly changed in the current repo. The broader discussion is now here:

* https://github.com/frequenz-floss/frequenz-client-base-python/discussions/41

Reply:
>I would just make the class iterable directly (also Iterator[MetricSample][1](https://github.com/frequenz-floss/frequenz-client-reporting-python/pull/16/files#user-content-fn-1-bbcbfbb94d5c423a8a52d25024597561) is a shorter and more idiomatic alias for Generator[MetricSample, None, None][2](https://github.com/frequenz-floss/frequenz-client-reporting-python/pull/16/files#user-content-fn-2-bbcbfbb94d5c423a8a52d25024597561))

Agree, will change this. In general a generator was used for lazy iteration, but that doesn't need to be reflected in the type annotation.

Reply:
Iterators also provides *lazy iteration*. Generators are just more powerful as they also let you pass messages between the iterator and the outside world.

Reply:
https://github.com/frequenz-floss/frequenz-client-reporting-python/pull/36